Rabbitmq Support

taybin opened the pull request:
Could use some more integration tests, and maybe some retry logic if the connection dies.

diranged commented on transport/transporters/rabbitmq/config.go:
```diff
@@ -0,0 +1,74 @@
+	altsrc.NewStringFlag(cli.StringFlag{
+		Name:   ConfVarExchangeName,
+		Usage:  "RabbitMQ Exchange Name",
+		EnvVar: "BIFROST_RABBITMQ_EXCHANGE_NAME",
```
Are there more standard RabbitMQ variables we can use here - rather than Bifrost-specific ones? As an example, for Kinesis we use the standard AWS variables (https://github.com/Nextdoor/pg-bifrost/blob/cabf3e6df5d33a7a7a719adcf6b12684311b869f/transport/transporters/kinesis/config.go#L46).

diranged replied:
It looks like the AMQP package supports `AMQP_URL`. Perhaps it would be better to simply do that, than have all these different variables?

diranged replied:
https://github.com/streadway/amqp/blob/e5adc2ada8b8efff032bf61173a233d143e9318e/doc.go#L73

taybin replied:
We could do that.

diranged commented on transport/transporters/rabbitmq/config.go:
```diff
@@ -0,0 +1,74 @@
+)
+
+var Flags = []cli.Flag{
+	cli.StringFlag{
```
Can you add in SSL support?

taybin replied:
Sure.

taybin replied:
In-depth TLS support is a little more difficult (see example: https://godoc.org/github.com/streadway/amqp#DialTLS), but just turning it on should be as easy as setting the schema to `amqps`.

nehalrp commented on transport/transporters/rabbitmq/factory.go:
```diff
@@ -0,0 +1,200 @@
+
+	conf := getConfig(transportConfig)
+
+	conn := createConnection(conf.host, conf.username, conf.password, conf.vhost, log)
```
With the Connection-Channel model in RabbitMQ, how are network failures and reconnects handled on the TCP level in the `wabbit` library? And how are we handling those particular errors that propagate up? Since it looks like all of our Transport threads will be sharing a single connection.

taybin replied:
That is the normal model with RabbitMQ.  One connection and then each reader/writer opens a channel inside that connection.

The only time that is not suggested is if you want to avoid TCP pushback on publishing affecting the ability to consume messages.  That's not the case here where we are only publishing.  (https://derickbailey.com/2014/03/26/2-lessons-learned-and-3-resources-for-for-learning-rabbitmq-on-nodejs/)

taybin replied:
I see what you mean though.  I've added a shared ConnectionManager that listens for connection failures and will create a new connection for all the transports to create new channels from as needed.

nehalrp commented on transport/transporters/rabbitmq/factory.go:
```diff
@@ -0,0 +1,200 @@
+			"noWait":   false,
+		},
+	)
+	log.WithField("exchangeName", conf.exchange).WithError(err).Fatal("Could not declare exchange")
```
Should this be wrapped in an `err != nil` conditional?

taybin replied:
Yes, I have that fixed locally and will be adding test coverage to catch similar errors in the future.

nehalrp commented on transport/transporters/rabbitmq/factory.go:
```diff
@@ -0,0 +1,200 @@
+	return conn
+}
+
+func connectionCleanup(shutdownHandler shutdown.ShutdownHandler, conn wabbit.Conn, log *logrus.Entry) {
```
Can you add some comments for this func?

taybin replied:
Will do.

taybin replied:
Ended up deleting that function entirely.

taybin commented on main/main.go:
```diff
@@ -34,12 +34,13 @@ import (
 	"github.com/jackc/pgx"
 	"github.com/sirupsen/logrus"
-	"gopkg.in/Nextdoor/cli.v1"
+	cli "gopkg.in/Nextdoor/cli.v1"
```
My editor/formatter did that automatically.  It might have been confused by the `v1` portion.

taybin replied:
FWIW, I tracked this change down to `goreturns` changing the alias of the import, somehow.  I'll revert that change if you like.

stlava commented on transport/transporters/rabbitmq/transporter/transporter.go:
```diff
@@ -0,0 +1,235 @@
+	var err error
+	op := operation{}
+	for _, msg := range messagesSlice {
+		err = json.Unmarshal(msg.Json, &op)
```
`MarshalledMessage` already has `operation`. It's reasonable to add `table` too so you don't have to unmarshal here.

taybin replied:
Oh, perfect.  Thank you.

taybin commented:
I believe this is in a better state for merging, if anyone wants to take a look at it again, especially the logic around retrying connections.

taybin commented:
I believe this is ready to be looked at again.

- The configuration has been simplified to a single envvar
- It supports the amqps TLS protocol
- Reverted the `cli.v1` aliasing change
- it will retry creating new connections if the connection should close, which will be shared amongst the transporters
- The json unmarshalling inside the busy loop has been removed.
- Added unittests

Please let me know of any other feedback.

stlava commented on transport/transporters/rabbitmq/transporter/connection.go:
```diff
@@ -0,0 +1,108 @@
+			return nil
+		}
+
+		err := backoff.Retry(operation, backoff.WithContext(cm.retryPolicy, ctx))
```
I suggest adding a reset on the retry policy.
```
defer func() {
	cm.retryPolicy.Reset()
}()
```

stlava commented on transport/transporters/rabbitmq/transporter/connection.go:
```diff
@@ -0,0 +1,108 @@
+	"github.com/sirupsen/logrus"
+)
+
+//go:generate mockgen -destination=mock_transporter.go -package=mocks github.com/Nextdoor/pg-bifrost.git/transport/transporters/rabbitmq/transporter ConnectionGetter
```
Need to change destination to `mocks/mock_transporter.go` otherwise it dumps it into this same directory. Also please move this to `transporter.go`.

stlava commented on transport/transporters/rabbitmq/factory.go:
```diff
@@ -0,0 +1,133 @@
+	logger.SetLevel(logrus.InfoLevel)
+}
+
+func makeDialer(connStr string) DialerFn {
```
I'm getting a `transport/transporters/rabbitmq/factory.go:43:33: undefined: DialerFn` when running `go list ./...  | xargs go vet`

taybin commented:
- Fixed the DialerFn `govet` error
- Changed the connection manager's mock destination
- reset the connection manager retry policy after use

stlava commented on transport/transporters/rabbitmq/transporter/transporter.go:
```diff
@@ -0,0 +1,261 @@
+			return
+		}
+
+		confirms, err = t.waitForConfirmations(closeNotify, uint64(len(messagesSlice)), confirms)
```
The transport doesn't have retry logic for attempting to publish a message or a batch of messages again. Is this something the underlying library does?

taybin replied:
My assumption here was that a failure in delivery would be caused by a disconnection, which is handled by the transport layer by exiting `transport()` and retrying with a new connection and a new call to `transport()` in the for-loop in `StartTransporting()`.

taybin replied:
Do you think additional retry logic is appropriate here?  There are other ways a call could fail, but they wouldn't get resolved by a retry, and I believe I've avoided them here.  (Turning on the mandatory flag, so that every message ends up in a queue is one example.  For my use case, it's okay if there isn't a queue, and it's easy enough for the user to create a queue for the messages anyways.)

stlava replied:
If there is any failure in `transport` then the batch would have already been read off of the `inputChan`. We would wind up in a scenario where there is a gap because the batch was consumed but never retried or acknowledged. This case has to be covered.

There's obviously a handful of ways to do this. I recommend the transport function calling out to a get connection function which would create a new or return an existing connection. Something like this:

```
transportWithRetry(context, batch) error {
   conn, err = getConn()
   if err... return err (this would cause retrier to retry)

  sendMessages(... batch)
  waitForConfirmations(... batch)
}

getConn() {
   if conn is closed
        return new conn, err
   else
       return conn, nil
}

StartTransporting() {
   for {
      read batch / term context

      retry(transportWithRetry)
   }
}
```

taybin replied:
I see what you're saying.  I'll address this.

taybin replied:
I've brought the rabbitmq transporter much closer to the kinesis transporter's retry handling.  I believe it is now identical, or as close as possible, in behavior.

stlava commented on transport/transporters/rabbitmq/config.go:
```diff
@@ -0,0 +1,52 @@
+	altsrc.NewStringFlag(cli.StringFlag{
+		Name:   ConfVarURL,
+		Usage:  "RabbitMQ URL",
+		EnvVar: "BIFROST_RABBITMQ_URL",
```
@diranged what are your thoughts about prefixed transport config vars? Looks like `RABBITMQ_URL` is a standard URL but `RABBITMQ_EXCHANGE_NAME` isn't. And certainly neither is `RABBITMQ_BATCH_SIZE`. Question is do we prefix everything, nothing, or anything custom (non-standard). Note we do similar in the kinesis transport with aws vars and stream name.

diranged replied:
Re `RABBITMQ_BATCH_SIZE` - couldn't we do someting more generic that spans different transporters, like `BATCH_SIZE`?

Re the Exchange name.. is that the same as the Virtual Host thats part of the URI specification: https://www.rabbitmq.com/uri-spec.html ?

taybin replied:
The exchange name is completely different from the virtual host and it wouldn't be appropriate to include it in the URI.  I could, but I've never seen any other RabbitMQ client do so.

diranged replied:
@taybin,
  Fair enough - I was just checking. I am fine with us adding variables that don't have defined standards. Just want to try to use standards where we can.

stlava commented on transport/transporters/rabbitmq/transporter/transporter.go:
```diff
@@ -0,0 +1,346 @@
+		}
+
+		if remaining > 0 {
+			messagesSlice = messagesSlice[uint64(len(messagesSlice))-remaining:]
```
I'm not very familiar with rabbitmq (and the library), are we guaranteed to have messages sent and acknowledged in order? Ie is it safe to assume we only need to retry `uint64(len(messagesSlice))-remaining:`?

taybin replied:
Rereading the example code, no.  This isn't a safe assumption and I need to redo this to use a map of deliveryTags to messages.

taybin replied:
Just how the golang client behaves here is a little confusing.  Hopefully the authors can explain it better for me.  https://github.com/streadway/amqp/issues/396

taybin replied:
It is safe to assume we will receive the acknowledgements in order, as the golang library will reorder the acknowledgements for us.  This does increase the possibility of resubmitting an already delivered message in the event of a crash, but since this is already a possibility, the user needs to be okay with that already.

With that in mind, I think the above code is correct.

stlava commented on transport/transporters/rabbitmq/transporter/transporter.go:
```diff
@@ -0,0 +1,346 @@
+		remaining = desiredCount - t.channelConfirms
+		select {
+		case confirm := <-t.publishNotify:
+			t.log.Info("received confirmation")
```
This should probably be a trace log message or removed.

stlava commented on transport/transporters/rabbitmq/transporter/transporter.go:
```diff
@@ -0,0 +1,346 @@
+
+		go t.closeHandler(ctx, t.closeNotify)
+
+		t.log.Info("Created new channel")
```
This should probably be a debug level log.

stlava commented:
@taybin this looks great. Thanks for making the changes. I left a few minor comments for you. Additionally can you please add a test to cover when `waitForConfirmations` returns an error so that the retry logic is covered.

taybin commented:
@stlava Will do.  I also need to redo the confirmation handler to use a map instead of a slice for out of order confirmations.

taybin commented:
Changed log levels as requested.

After checking with the golang library maintainers, I believe the current use of a slice to handle tracking what messages were delivered to RabbitMQ and what need to be resent is the best solution.  It is possible in the event of a crash, either by RabbitMQ or pg-bifrost, that some successfully delivered messages will be resent, but the code as is should not lose any that were not acknowledged by RabbitMQ.

taybin commented:
Still need to add a test for when `waitForConfirmations` returns an error, as requested.

taybin commented:
@stlava I've added the test on `waitForConfirmation()` returning an error.  I don't believe there are anymore outstanding requests.

taybin commented:
FWIW, we've deployed this and it's working great.  That is due to the resiliency of your design.  Thank you for open sourcing this.

stlava commented:
Thanks for contributing! Will merge this in and do a release shortly.
